lib/index.d.ts

```diff
@@ -0,0 +1,161 @@
+import ascend from "./ascend";
+import az from "./az";
+import both from "./both";
+import { call } from "when/node";
+
+declare let nanoutils: Nanoutils.Static
+
+declare namespace Nanoutils {
+  type Predicate1<T1> = (value: T1) => boolean
+
+  // addIndex
+  interface ArrayIterator<T1, T2> {
+    (change: (value: T1) => T2, array: ReadonlyArray<T1>): T2[]
+  }
+
+  interface IndexReducer<T1, T2> {
+    (value: T1, index: number): T2
+  }
+
+  interface AddIndexCurried2<T1, T2> {
+    (reducer: IndexReducer<T1, T2>): (array: ReadonlyArray<T1>) => T2[]
+    (reducer: IndexReducer<T1, T2>, array: ReadonlyArray<T1>): T2[]
+  }
+
+  // adjust
+  interface AdjustCurried2<T1, T2> {
+    (index: number): (array: ReadonlyArray<T1>) => T2[]
+    (index: number, array: ReadonlyArray<T1>): T2[]
+  }
+
+  // adjustIn
+  interface AdjustInCurried2<T1, T2> {
+    (predicate: Predicate1<T1>): (array: ReadonlyArray<T1>) => T2[]
+    (predicate: Predicate1<T1>, array: ReadonlyArray<T1>): T2[]
+  }
+
+  // applySpec
+  interface ApplySpecObject<T1> {
+    [key: string]: T1
+    [key: string]: ApplySpecObject<T1>
+  }
+
+  // ascend
+  interface Comparator<T1> {
+    (first: T1, second: T1): -1 | 0 | 1
+  }
+
+  // assocPath
+  interface AssocPathObject<T1> {
+    [key: string]: T1
+    [key: string]: AssocPathObject<T1>
+  }
+
+  // call
+  interface CallFunction<T1, T2> {
+    (): CallFunction<T1, T2>
+    (...array: T1[]): T2
+  }
+
+  interface Static {
+    // add
+    add(first: number): (second: number) => number
+    add(first: number, second: number): number
+
+    // addIndex
+    addIndex<T1, T2>(iterator: ArrayIterator<T1, T2>): AddIndexCurried2<T1, T2>
+    addIndex<T1, T2>(iterator: ArrayIterator<T1, T2>, reducer: IndexReducer<T1, T2>): (array: ReadonlyArray<T1>) => T2[]
+    addIndex<T1, T2>(iterator: ArrayIterator<T1, T2>, reducer: IndexReducer<T1, T2>, array: ReadonlyArray<T1>): T2[]
+
+    // adjust
+    adjust<T1, T2>(change: (value: T1) => T2): AdjustCurried2<T1, T2>
+    adjust<T1, T2>(change: (value: T1) => T2, index: number): (array: ReadonlyArray<T1>) => (T1 | T2)[]
+    adjust<T1, T2>(change: (value: T1) => T2, index: number, array: ReadonlyArray<T1>): (T1 | T2)[]
+
+    // adjustIn
+    adjust<T1, T2>(change: (value: T1) => T2): AdjustInCurried2<T1, T2>
+    adjust<T1, T2>(change: (value: T1) => T2, predicate: Predicate1<T1>): (array: ReadonlyArray<T1>) => T2[]
+    adjust<T1, T2>(change: (value: T1) => T2, predicate: Predicate1<T1>, array: ReadonlyArray<T1>): T2[]
+
+    // all
+    all<T1>(predicate: Predicate<T1>): (predicate: Predicate1<T1>) => boolean
+    all<T1>(predicate: Predicate<T1>, predicate: Predicate1<T1>): boolean
```

Comment:
something wrong with names of parameters, need to change

Reply:
also a typo, `Predicate` is not defined its `Predicate1`